Stealth Transactions

cpacia opened the pull request.

cpacia committed:
Merge branch 'master' of https://github.com/OpenBazaar/openbazaar-go

cpacia committed:
Merge branch 'master' of https://github.com/OpenBazaar/openbazaar-go

cpacia committed:
Add import key functionality

cpacia committed:
Add stealth transaction feature

New feature allows bitcoin payments to be sent to an OpenBazaar peerId or handle.
When this type of payment is made the recipient's master public key is looked
up and a chaincode is generated using ECDH with an ephemeral key pair. The coins
get sent to a seemingly random address, not linked to the master public key.
The ephemeral public key is included in op_return and the recipient can derive
the key needed to spend the coins from the same ECDH operation. The transaction
is relayed by the remote peers by pushing a flag byte to op_return which is also
set in the filter.

cpacia committed:
Update spvwallet dependency

jackkleeman commented:
Is this aimed at vendors withdrawing money? Very cool.

cpacia commented:
It allows someone to send bitcoins to another user's peerID. The coins arrive at a different bitcoin address each time preventing address reuse.

cpacia committed:
Fix bug in bitcoind stealth impl

jackkleeman commented:
Will this be used as part of the protocol or is it more of an extra useful feature?

cpacia commented:
It's only going to be used in the wallet as far as I know.

cpacia committed:
Fix bitcoind stealth rpc calls

cpacia committed:
Fix keys test

cpacia committed:
Disable stealth receive for bitcoind impl

cpacia committed:
Shutdown bitcoind properly

cpacia committed:
Fix getbalance for bitcoind impl

JustinDrake commented:
Can stealth transactions be used for direct OpenBazaar payments (i.e. no moderator)?

cpacia commented:
Given that not all wallet implementations can handle stealth transactions (for example bitcoind can't) it's probably not a good idea.

cpacia committed:
Remove unused fields from bitcoind NotificationListener

cpacia committed:
Merge branch 'master' of https://github.com/OpenBazaar/openbazaar-go

cpacia committed:
Fix bug shutting down bitcoind impl

In the resync method the rpcclient shutdown function wasn't shutting down bitcoind.
This commit changes the rpc call to use one that will work. Also increase the
timeout time on startup.

tyler-smith commented on repo/db/keys.go:
```diff
@@ -92,6 +111,25 @@ func (k *KeysDB) GetPathForScript(scriptPubKey []byte) (spvwallet.KeyPath, error
+	k.lock.Lock()
+	defer k.lock.Unlock()
+
+	stmt, err := k.db.Prepare("select key from keys where scriptPubKey=? and purpose=-1")
```
We should catch this to avoid a nil pointer panic on 121

cpacia committed:
Catch error in key db
